### PR TITLE
Retry all servers on RPC call failure

### PR DESCRIPTION
**Pull request by schmichael**
rpcproxy is refactored into serverlist which prioritizes good servers
over servers in a remote DC or who have had a failure.

Registration, heartbeating, and alloc status updating will retry faster
when new servers are discovered.

Consul discovery will be retried more quickly when no servers are
available (eg on startup or an outage).

### CODE REVIEW

File: client/client.go [hunk 1]
```diff
@@ -1396,169 +1478,175 @@ func (c *Client) deriveToken(alloc *structs.Allocation, taskNames []string, vcli
...
+func (c *Client) triggerDiscovery() {
+	select {
+	case <-c.doDisco:
+		// Discovery goroutine was released to execute
+	default:
+		// Discovery goroutine was already running
+	}
+}
+
+func (c *Client) consulDiscovery() {
```
Comment (dadgar) [inline client/client.go]: Comment the method

File: client/client.go [hunk 2]
```diff
@@ -1396,169 +1478,175 @@ func (c *Client) deriveToken(alloc *structs.Allocation, taskNames []string, vcli
...
-		// If the last heartbeat didn't contain a leader, give the
-		// Nomad server this Agent is talking to one more attempt at
-		// providing a heartbeat that does contain a leader.
-		if atomic.LoadInt32(&c.lastHeartbeatFromQuorum) == 1 && now.Before(c.consulPullHeartbeatDeadline) {
-			c.heartbeatLock.Unlock()
-			return nil
+// triggerDiscovery causes a consul discovery to begin (if one hasn't alread)
+func (c *Client) triggerDiscovery() {
+	select {
+	case <-c.doDisco:
```
Comment (dadgar) [inline client/client.go]: Change to be: `case c.doDisco <- struct{}{}`

File: client/client.go [hunk 3]
```diff
@@ -1396,169 +1478,175 @@ func (c *Client) deriveToken(alloc *structs.Allocation, taskNames []string, vcli
...
+		// Discovery goroutine was released to execute
+	default:
+		// Discovery goroutine was already running
+	}
+}
+
+func (c *Client) consulDiscovery() {
+	for {
+		select {
+		case c.doDisco <- struct{}{}:
```
Comment (dadgar) [inline client/client.go]: Change to be `case <-c.doDisco:`

File: client/client.go [hunk 4]
```diff
@@ -1396,169 +1478,175 @@ func (c *Client) deriveToken(alloc *structs.Allocation, taskNames []string, vcli
...
+				c.logger.Printf("[ERR] client.consul: error discovering nomad servers: %v", err)
+			}
+		case <-c.shutdownCh:
+			return
 		}
-		c.heartbeatLock.Unlock()
+	}
+}
+
+func (c *Client) doConsulDisco() error {
```
Comment (dadgar) [inline client/client.go]: Comment plus change name to be `consulDiscoveryImpl`

File: client/client.go [hunk 5]
```diff
@@ -1396,169 +1478,175 @@ func (c *Client) deriveToken(alloc *structs.Allocation, taskNames []string, vcli
...
-			// considering all known Nomad servers and
-			// eventually, hopefully, find their way to a Nomad
-			// Server that has quorum (assuming Consul has a
-			// server list that is in the majority).
-			for _, s := range nomadServerServices {
-				c.rpcProxy.AddPrimaryServer(s)
+	// Notify waiting rpc calls. Wait briefly in case initial rpc
+	// just failed but the calling goroutine isn't selecting on
+	// discovered yet.
+	const dur = 50 * time.Millisecond
```
Comment (dadgar) [inline client/client.go]: There is always the case that they will be on that border. We can remove this

File: client/client.go [hunk 6]
```diff
@@ -1396,169 +1478,175 @@ func (c *Client) deriveToken(alloc *structs.Allocation, taskNames []string, vcli
...
-			ar.taskStatusLock.RLock()
-			taskStates := copyTaskStates(ar.taskStates)
-			ar.taskStatusLock.RUnlock()
-			for taskName, taskState := range taskStates {
-				// Only keep running tasks
-				if taskState.State == structs.TaskStateRunning {
-					d := consul.NewExecutorDomain(allocID, taskName)
-					domains = append(domains, d)
-				}
+func (c *Client) consulReaper() {
```
Comment (dadgar) [inline client/client.go]: Comment on method

I would prefer you initialize the ticker here

File: client/client.go [hunk 7]
```diff
@@ -1396,169 +1478,175 @@ func (c *Client) deriveToken(alloc *structs.Allocation, taskNames []string, vcli
...
+			return
 		}
-
-		return c.consulSyncer.ReapUnmatched(domains)
 	}
-	if c.config.ConsulConfig.AutoAdvertise {
-		c.consulSyncer.AddPeriodicHandler("Nomad Client Services Sync Handler", consulServicesReaperFn)
+}
+
+func (c *Client) doConsulReap() error {
```
Comment (dadgar) [inline client/client.go]: Comments and rename to `consulReapImpl`

File: client/client.go [hunk 8]
```diff
@@ -386,8 +398,8 @@ func (c *Client) Stats() map[string]map[string]string {
 	stats := map[string]map[string]string{
 		"client": map[string]string{
 			"node_id":         c.Node().ID,
-			"known_servers":   toString(uint64(c.rpcProxy.NumServers())),
-			"num_allocations": toString(uint64(numAllocs)),
+			"known_servers":   strconv.Itoa(len(c.servers.all())),
```
Comment (schmichael) [inline client/client.go]: Replace the existing count or add a new entry?

Comment (schmichael) [inline client/client.go]: (I'd vote for replace existing, but I don't know if people rely on this to be stable between releases)

Comment (dadgar) [inline client/client.go]: I would replace

File: client/client.go (hunk 2 above)
Comment (schmichael) [inline client/client.go]: I have no idea why I inverted the usual chan flow... fixing.

File: client/client.go [hunk 9]
```diff
@@ -44,6 +43,10 @@ const (
 	// datacenters looking for the Nomad server service.
 	datacenterQueryLimit = 9
 
+	// consulReaperIntv is the interval at which the consul reaper will
```
Comment (schmichael) [inline client/client.go]: Fixed everywhere (I think)

File: client/client.go [hunk 10]
```diff
@@ -105,26 +108,22 @@ type Client struct {
...
+	// servers is the (optionally prioritized) list of nomad servers
+	servers *serverlist
+
+	// heartbeat related times for tracking how often to heartbeat
+	lastHeartbeat time.Time
+	heartbeatTTL  time.Duration
+	heartbeatLock sync.Mutex
+
+	// doDisco triggers consul discovery; see triggerDiscovery
+	doDisco chan struct{}
```
Comment (schmichael) [inline client/client.go]: Fixed! (Much clearer!)

File: client/client.go [hunk 11]
```diff
@@ -154,14 +156,20 @@ type Client struct {
 	vaultClient vaultclient.VaultClient
 }
 
+var noServers = errors.New("no servers")
```
Comment (schmichael) [inline client/client.go]: Fixed

File: client/client.go [hunk 12]
```diff
@@ -202,17 +210,25 @@ func NewClient(cfg *config.Config, consulSyncer *consul.Syncer, logger *log.Logg
 	// Create the RPC Proxy and bootstrap with the preconfigured list of
 	// static servers
 	c.configLock.RLock()
-	c.rpcProxy = rpcproxy.NewRPCProxy(c.logger, c.shutdownCh, c, c.connPool)
-	for _, serverAddr := range c.configCopy.Servers {
-		c.rpcProxy.AddPrimaryServer(serverAddr)
+	if len(c.configCopy.Servers) > 0 {
+		if err := c.SetServers(c.configCopy.Servers); err != nil {
+			logger.Printf("[WARN] None of the configured servers are valid: %v", err)
```
Comment (schmichael) [inline client/client.go]: Fixed

File: client/client.go [hunk 13]
```diff
@@ -350,33 +356,39 @@ func (c *Client) Shutdown() error {
...
+			c.logger.Printf("[DEBUG] client: %v", errmsg)
+			c.servers.failed(s)
+			continue
+		}
+		c.servers.good(s)
+		return nil
 	}
-	return nil
+
+	// Force consul discovery ASAP since we have no healthy servers
```
Comment (schmichael) [inline client/client.go]: Removed.

It does not always force discovery. This line described a now deleted piece of code.

Currently heartbeating is the only RPC caller that immediately causes discovery on *any* error.

File: client/client.go (hunk 8 above)
Comment (schmichael) [inline client/client.go]: Fixed

File: client/client.go (hunk 1 above)
Comment (schmichael) [inline client/client.go]: Fixed

File: client/client.go (hunk 4 above)
Comment (schmichael) [inline client/client.go]: Fixed

File: client/client.go (hunk 5 above)
Comment (schmichael) [inline client/client.go]: Removed.

I think this was a clumsy workaround to get around startup races early on that AFAICT (both looking at the code again and testing it manually) is no longer necessary.

File: client/client.go (hunk 6 above)
Comment (schmichael) [inline client/client.go]: Fixed.

Good call; there was no need for the ticker to be on the struct at all!

File: client/client.go (hunk 7 above)
Comment (schmichael) [inline client/client.go]: Fixed

File: client/client.go [hunk 14]
```diff
@@ -892,9 +948,6 @@ func (c *Client) registerNode() error {
 	}
 	var resp structs.NodeUpdateResponse
 	if err := c.RPC("Node.Register", &req, &resp); err != nil {
-		if time.Since(c.start) > registerErrGrace {
```
Comment (dadgar) [inline client/client.go]: Why did you remove this?

Comment (schmichael) [inline client/client.go]: It merely wrapped an error with a tiny bit more context in certain circumstances.

Instead I just added that tiny bit more context to the log line that handles this error, so there's no need to wrap it: https://github.com/hashicorp/nomad/pull/1735/files/ec81fc9baf776efb30bd752f64688ad562d915c6#diff-1cd41a2611da49ce63630d5a51bd82c5L876

It also kind of seemed like dead code since the error wrapping only affected logging. It seemed like maybe once upon a time it affected retry logic, but it doesn't  anymore on master or this PR.

Comment (dadgar) [inline client/client.go]: If you look at the comment on the `registerErrGrace`, it does have a point. It is to avoid error logs showing up in the `-dev` case. I would add it back

File: CHANGELOG.md [hunk 15]
```diff
@@ -4,6 +4,7 @@ IMPROVEMENTS:
   * core: Introduce node SecretID which can be used to minimize the available
     surface area of RPCs to malicious Nomad Clients [GH-1597] 
   * cli: `nomad alloc-status` shows allocation creation time [GH-1623]
+  * client: Retry all servers on RPC call failure [GH-1735]
```
Comment (dadgar) [inline CHANGELOG.md]: Failed RPCs are retried on all servers

File: CHANGELOG.md [hunk 16]
```diff
@@ -17,6 +18,8 @@ BUG FIXES:
   * client/fingerprint: Fix inconsistent CPU MHz fingerprinting [GH-1366]
   * discovery: Fix old services not getting removed from consul on update
     [GH-1668]
+  * discovery: Fix client flapping when server was in a different datacenter
```
Comment (dadgar) [inline CHANGELOG.md]: was -> is

+/- different datacenter as the client
